src/Workspaces/Core/Portable/Utilities/EditDistance.cs

```diff
@@ -1,175 +1,506 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Diagnostics;
+using static System.Math;
 
 namespace Roslyn.Utilities
 {
-    internal static class EditDistance
+    internal class EditDistance : IDisposable
     {
-        private const int MaxPooledArraySize = 256;
+        private string originalText;
+        private char[] originalTextArray;
+        private int threshold;
 
-        // Keep around a few arrays of size 256 that we can use for operations without
-        // causing lots of garbage to be created.  If we do compare items larger than
-        // that, then we will just allocate and release those arrays on demand.
-        private static ObjectPool<int[]> s_pool = new ObjectPool<int[]>(() => new int[MaxPooledArraySize]);
+        // Cache the result of the last call to IsCloseMatch.  We'll often be called with the same
+        // value multiple times in a row, so we can avoid expensive computation by returning the
+        // same value immediately.
+        private ValueTuple<string, bool, double> lastIsCloseMatchResult;
 
-        public static int[] GetArray(int size)
+        public EditDistance(string text, int? threshold = null)
         {
-            if (size <= MaxPooledArraySize)
+            originalText = text;
+            originalTextArray = ConvertToLowercaseArray(text);
+
+            // We only allow fairly close matches (in order to prevent too many
+            // spurious hits).  A reasonable heauristic for this is the Log_2(length) (rounded 
+            // down).  
+            //
+            // Strings length 1-3 : 1 edit allowed.
+            //         length 4-7 : 2 edits allowed.
+            //         length 8-15: 3 edits allowed.
```

Comment:
We have to be careful here.  First, the pattern matcher isn't cheap.  You can see how long NavigateTo takes when it has to search all symbols.  We need this code-action to run orders of magnitude faster.  Second, do we also handle spelling mistakes while also doing chunking?  (that sounds complex in terms of cost). Or do we just do the chunking approach if the spell check fails?  etc. etc.